src/main/scala/com/wavesplatform/matcher/market/BalanceWatcherWorkerActor.scala

```diff
@@ -0,0 +1,111 @@
+  private type ChangesByAddress = Map[String, Portfolio]
+
+  private var requestIdCounter = 0L
+
+  override def receive: Receive = {
+    case BalanceChanged(changes) => becomeWorking(changes)
+  }
+
+  private def waitOrders(taskId: Long,
+                         stashedChanges: ChangesByAddress,
+                         changesByAddress: ChangesByAddress,
+                         waitOrdersTimeout: Cancellable): Receive = {
+    case GetActiveOrdersByAddressResponse(`taskId`, address, orders) =>
+      ordersToDelete(changesByAddress, address, orders)
+        .foreach {
+          case (pair, id) =>
+            orderHistory ! ForceCancelOrderFromHistory(id)
+            matcher ! ForceCancelOrder(pair, id)
+        }
+
+      val updated = changesByAddress - address
+      if (updated.isEmpty) {
+        if (stashedChanges.isEmpty) context.become(receive)
+        else becomeWorking(stashedChanges)
+      } else context.become(waitOrders(taskId, stashedChanges, updated, reschedule(waitOrdersTimeout)))
+
+    case x: BalanceChanged =>
+      context.become(
+        waitOrders(
+          taskId = taskId,
+          stashedChanges = replaceChanges(stashedChanges, x.changesByAddress),
+          changesByAddress = changesByAddress,
+          waitOrdersTimeout = reschedule(waitOrdersTimeout)
+        ))
+
+    case _: ReceiveTimeout =>
+      log.warn(s"Timeout to process orders for ${changesByAddress.size} addresses has been reached")
+      if (stashedChanges.isEmpty) context.become(receive)
+      else becomeWorking(stashedChanges)
+  }
+
+  private def becomeWorking(changes: ChangesByAddress): Unit = {
+    changes.keys
+      .map(GetActiveOrdersByAddress(requestIdCounter, _))
+      .foreach(orderHistory ! _)
+
+    context.become(waitOrders(requestIdCounter, Map.empty, changes, reschedule(EmptyCancellable)))
+    requestIdCounter += 1
+  }
+
+  private def reschedule(old: Cancellable): Cancellable = {
+    import context.dispatcher
+
+    old.cancel()
+    context.system.scheduler.scheduleOnce(TimeoutToProcessChanges, self, ReceiveTimeout)
+  }
+
+  private def replaceChanges(orig: ChangesByAddress, replacement: ChangesByAddress): ChangesByAddress = orig ++ replacement
+
+  private def ordersToDelete(portfolios: Map[String, Portfolio], ownerAddress: String, orders: Seq[LimitOrder]): OrdersToDelete = {
+    portfolios
+      .get(ownerAddress)
+      .map { portfolio =>
+        val ordersByPriority = orders.sortBy(_.order.timestamp)(Ordering[Long].reverse)
+        val (_, r) = ordersByPriority.foldLeft((portfolio, List.empty: OrdersToDelete)) {
+          case ((restPortfolio, toDelete), limitOrder) =>
+            val updatedPortfolio = restPortfolio
+              .copy(balance = restPortfolio.balance - restPortfolio.leaseInfo.leaseOut)
+              .remove(limitOrder.spentAcc.assetId, limitOrder.getSpendAmount)
+              .flatMap(_.remove(None, limitOrder.remainingFee))
```

Comment:
I guess orders which buy Waves for BTC wouldn't work here if there are not enough waves for a fee.

Reply:
👍 I'll create a test and fix this